File: src/limitedwip/autorevert/components/AutoRevertComponent.java
package limitedwip.autorevert.components;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.components.AbstractProjectComponent;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.project.Project;
import limitedwip.autorevert.AutoRevert;
import limitedwip.autorevert.ui.IdeActions2;
import limitedwip.autorevert.ui.IdeNotifications2;
import limitedwip.common.settings.LimitedWIPSettings;
import limitedwip.common.LimitedWipCheckin;
import limitedwip.common.settings.LimitedWipConfigurable;
import limitedwip.common.TimerComponent;

public class AutoRevertComponent extends AbstractProjectComponent implements LimitedWipConfigurable.Listener, LimitedWipCheckin.Listener {
	private final TimerComponent timer;
	private AutoRevert autoRevert;

	protected AutoRevertComponent(Project project) {
		super(project);
		timer = ApplicationManager.getApplication().getComponent(TimerComponent.class);
	}

	@Override public void projectOpened() {
		LimitedWIPSettings settings = ServiceManager.getService(LimitedWIPSettings.class);
		IdeNotifications2 ideNotifications = new IdeNotifications2(myProject);
		autoRevert = new AutoRevert(ideNotifications, new IdeActions2(myProject)).init(convert(settings));

		timer.addListener(new TimerComponent.Listener() {
			@Override public void onUpdate(int seconds) {
				autoRevert.onTimer(seconds);
			}
		}, myProject);

		LimitedWipConfigurable.registerSettingsListener(myProject, this);
		LimitedWipCheckin.registerListener(myProject, this);
	}

	@Override public void onSettingsUpdate(LimitedWIPSettings settings) {
		autoRevert.onSettings(convert(settings));
	}

	public void startAutoRevert() {
		autoRevert.start();
	}

	public boolean isAutoRevertStarted() {
		return autoRevert.isStarted();
	}

	public void stopAutoRevert() {
		autoRevert.stop();
	}

	public void onSuccessfulCheckin(boolean allFileAreCommitted) {
		if (allFileAreCommitted) {
			autoRevert.onAllFilesCommitted();
		}
	}

	private static AutoRevert.Settings convert(LimitedWIPSettings settings) {
		return new AutoRevert.Settings(
				settings.autoRevertEnabled,
				settings.secondsTillRevert(),
				settings.notifyOnRevert
		);
	}
}

File: src/limitedwip/autorevert/components/AutorevertComponent.kt
package limitedwip.autorevert.components

import com.intellij.openapi.application.ModalityState
import com.intellij.openapi.project.Project
import com.intellij.openapi.startup.StartupActivity
import limitedwip.autorevert.AutoRevert
import limitedwip.autorevert.ui.AutoRevertStatusBarWidget
import limitedwip.common.TimerAppComponent
import limitedwip.common.settings.LimitedWipSettings
import limitedwip.common.vcs.SuccessfulCheckin
import limitedwip.common.vcs.defaultChangeList
import limitedwip.common.vcs.invokeLater

class AutoRevertComponentStartup: StartupActivity {
    override fun runActivity(project: Project) = AutoRevertComponent(project).start()
}

class AutoRevertComponent(private val project: Project) {
    @Volatile private var enabled = false

    fun start() {
        val settings = LimitedWipSettings.getInstance(project).toAutoRevertSettings()
        val widget = AutoRevertStatusBarWidget(project)
        val autoRevert = AutoRevert(Ide(project, settings, widget), settings)
        widget.onClick {
            autoRevert.onPause()
        }

        enabled = settings.enabled

        TimerAppComponent.getInstance().addListener(project, object: TimerAppComponent.Listener {
            override fun onUpdate() {
                // Optimisation to avoid scheduling task when component is not enabled.
                if (!enabled) return

                invokeLater(ModalityState.any()) {
                    // Project can be closed (disposed) during handover between timer thread and EDT.
                    if (project.isDisposed) return@invokeLater
                    autoRevert.onTimer(hasChanges = project.defaultChangeList()?.changes?.isNotEmpty() ?: false)
                }
            }
        })

        val rollbackListener = RollbackListener(project) { allChangesRolledBack ->
            if (allChangesRolledBack) autoRevert.onAllChangesRolledBack()
        }

        LimitedWipSettings.getInstance(project).addListener(project, object: LimitedWipSettings.Listener {
            override fun onUpdate(settings: LimitedWipSettings) {
                autoRevert.onSettingsUpdate(settings.toAutoRevertSettings())
                enabled = settings.autoRevertEnabled
                if (enabled) rollbackListener.enable() else rollbackListener.disable()
            }
        })

        SuccessfulCheckin.registerListener(project, object: SuccessfulCheckin.Listener {
            override fun onSuccessfulCheckin(allChangesAreCommitted: Boolean) {
                if (allChangesAreCommitted) autoRevert.onAllChangesCommitted()
            }
        })
    }

    private fun LimitedWipSettings.toAutoRevertSettings() =
        AutoRevert.Settings(
            autoRevertEnabled,
            timeUnitTillRevert.toSeconds(timeTillRevert),
            notifyOnRevert
        )
}
